[REVIEW] Adding definition of done to wiki and updating README to link to it. [skip-ci] (#877)
rapidsai/cuml: wiki/DEFINITION_OF_DONE_CRITERIA.md

- **Documentation:** User-facing documentation should be complete and descriptive. Developer-facing documentation should be used for constructs which are complex and/or not immediately obvious. 

### C++

#### Design

- Existing prims are used wherever possible
- Array inputs and outputs to algorithms are accepted on device
- New prims created wherever there is potential for reuse across different algorithms or prims
- User-facing API is [stateless](cpp/DEVELOPER_GUIDE.md#public-cuml-interface) and follows the [plain-old data (POD)](https://en.wikipedia.org/wiki/Passive_data_structure) design paradigm
- Public API contains a C-Wrapper around the stateless API

> - Public API contains an Scikit-learn-like stateful wrapper around the stateless API

Reviewer: maybe mark these as optional?

#### Testing

- Prims: GTests with different inputs
- Algorithms: End-to-end GTests with different inputs and different datasets

#### Documentation

- Complete and comprehensive Doxygen strings explaining the public API, restrictions, and gotchas. Any array parameters should also note whether the underlying memory is host or device.
- Array inputs/outputs should also mention their expected size/dimension.
- If there are references to the underlying algorithm, they must be cited too.

### Python

#### Design

> - Python class is as "near drop-in replacement" for Scikit-learn (or relevant industry standard) API as possible

Reviewer 2: Expand a little - "parameters have the same names as scikit-learn, and where differences exist they are clearly documented in docstrings."

- It is recommended to open an initial PR with the API design if there are going to be significant differences with reference APIs, or lack of a reference API, to have a discussion about it. 

> - Python class is pickleable

Reviewer 2: Ensure that the test has been added to cuml/test/test_pickle.py

> - APIs use `input_to_dev_array` to accept flexible inputs and check their datatypes

Reviewer 3: ```suggestion
- APIs use `input_to_cuml_array` to accept flexible inputs and check their datatypes
```
Vintage function :P 

And perhaps add a line for use cumlarray?

Reviewer 2: Agreed - inherits from Base and uses CumlArray for inputs and outputs

#### Testing 

- Pytests for wrapper functionality against Scikit-learn using relevant datasets
- Stress tests against reasonable inputs (e.g short-wide, tall-narrow, different numerical precision)
- Pytests for pickle capability
- Pytests to evaluate correctness against Scikit-learn on a variety of datasets

> - Add algorithm to benchmarks packages

Reviewer 2: Add algorithm to benchmark package in `python/cuml/benchmarks/algorithms.py`

(just spelling it out for our noob friends!)

#### Documentation

> - Complete and comprehensive Pydoc strings explaining public API, restrictions, and gotchas

Reviewer 2: Maybe mention that they should follow numpydoc format?

- Docstrings include references to the underlying algorithm (e.g paper or Scikit-learn algorithm being implemented or a description of the algorithm used if nonstandard).

> ## Algorithm Status

Reviewer 2: Are we ready to launch this and classify all of the existing algorithms? (I know this is a broader discussion topic)

Reviewer 2: I would phrase this as - "We are beginning to assign a status of bronze, silver gold"

Reviewer 2: Honestly, I think we should hold back on bronze silver gold for now... Love the rest of the doc and think this is a great step, but I think this will be the most contentious section, so we should discus more.

Sometimes it is possible that only some of these requirements are satisfied before an algorithm gets released. We assign a status of `bronze`, `silver`, or `gold` to denote an algorithm's level of completeness and robustness.

Note that a change to an algorithm's classification can go up or down, depending on circumstances. For example, an algorithm that has been considered `Gold status` might drop to a `Silver status` when being updated to a new low-level feature or library. 

### Bronze: 
- Known design flaws affecting sustainability and/or performance require a significant effort / redesign / refactor

> - C++ checklist / API functionality is complete but Python API has not been built

Reviewer 2: I think if there is no python API at all we're not even bronze yet.

- Pure Python API exists without using C++ for processing when a C++ implementation would provide a significant performance boost

> - There are known issues that affect memory / accuracy / stability on some inputs

Reviewer 2: If it's a significant accuracy issue, I think we're sub-bronze... maybe stone? rust? But for memory I agree

> - Coverage of automated tests in C++ or Python is minimal and could be significantly improved

Reviewer 2: Instead of minimal maybe "insufficient"?

- Outdated / legacy code that requires significant work and refactoring

### Silver:
- Known design flaws affecting sustainability and/or performance are fairly minor.
- Automated tests in C++ or Python are not as extensive as they should be regarding different datasets / inputs, but they do test general functionality
- There are known (or potential) bugs that affect memory / accuracy, but only on a small subset of specific inputs
- There are known optimizations that would provide a significant performance boost over the current implementation

### Gold: 
- Design is robust, maintainable, and performant without any known flaws.

> - Algorithm is as performant as possible without any significant known or potential optimizations over the current implementation.

Reviewer 2: Not sure I agree with this one... how about "algorithm shows consistent speedups over CPU comparison and has no major known bottlenecks" -- I don't want to suggest that we have to optimize the hell out of everything before it's gold. If it's already quite fast and very _robust_ that's gold to me

> - Algorithm is consistently correct and bug free on all datasets.

Reviewer 2: "all tested dtasets"

- There are no known memory usage limitations resulting from the design choices that could be reasonably refactored.

> 

Reviewer 2: I would add a difficult requirement that algorithm has been used in other pipelines outside of cuml tests - customers or large scale demos and has been around for a bit of time. Happy to chat more to strategize phrasing, but I wouldn't want to declare something gold until it has been used by non-core team members even if we feel good about it

## Review Checklist

Aside from the general algorithm expectations outlined in the checklists above, code reviewers should use the following checklist to make sure the algorithm meets cuML standards. 

### All

- New files contain necessary license headers
- Diff does not contain files with excess formatting changes, without other changes also being made to the file

> - Code does not contain any immediately obvious memory leaks or garbage collection issues 

Reviewer 2: obvious -> "known, serious"

- **Performance:** Algorithms should be [benchmarked] and profiled regularly to spot potential bottlenecks, performance regressions, and memory problems.

### C++

#### Design

- Existing prims are used wherever possible
- Array inputs and outputs to algorithms are accepted on device
- New prims created wherever there is potential for reuse across different algorithms or prims
- User-facing API is [stateless](cpp/DEVELOPER_GUIDE.md#public-cuml-interface) and follows the [plain-old data (POD)](https://en.wikipedia.org/wiki/Passive_data_structure) design paradigm
- Public API contains a C-Wrapper around the stateless API
- (optional) Public API contains an Scikit-learn-like stateful wrapper around the stateless API

#### Testing

- Prims: GTests with different inputs
- Algorithms: End-to-end GTests with different inputs and different datasets

#### Documentation

- Complete and comprehensive [Doxygen](http://www.doxygen.nl/manual/docblocks.html) strings explaining the public API, restrictions, and gotchas. Any array parameters should also note whether the underlying memory is host or device.

> - Array inputs/outputs should also mention their expected size/dimension.

Reviewer 3: just a brief mention that they should also be explicit about whether they are device or host arrays (which sometimes is omitted in docs)

Author: This is mentioned in the sentence above. Would you prefer it to be on this line instead?

Reviewer 3: I am totally blind :P it is fine where it is

- Docstrings include references to any scientific papers or standard publications on the underlying algorithm (e.g paper or Scikit-learn algorithm being implemented or a description of the algorithm used if nonstandard).

## Review Checklist

Aside from the general algorithm expectations outlined in the checklists above, code reviewers should use the following checklist to make sure the algorithm meets cuML standards. 

### All

- New files contain necessary license headers
- Diff does not contain files with excess formatting changes, without other changes also being made to the file
- Code does not contain any known serious memory leaks or garbage collection issues
- Modifications are cohesive and in-scope for the PR's intended purpose
- Changes to the public API will not have a negative impact to existing users between minor versions (eg. large changes to very popular public APIs go through a deprecation cycle to preserve backwards compatibility)
- Where it is reasonable to do so, unexpected inputs fail gracefully and provide actionable feedback to the user
- Automated tests properly exercise the changes in the PR

> - New algorithms provide benchmarks (both C++ and Python) 

Reviewer 3: How about adding an item to see the list of printed slowest pytests in CI, to check that a new super slow test is not being introduced in the PR?

- **Documentation:** User-facing documentation should be complete and descriptive. Developer-facing documentation should be used for constructs which are complex and/or not immediately obvious. 

### C++

#### Design

- Existing prims are used wherever possible
- Array inputs and outputs to algorithms are accepted on device
- New prims created wherever there is potential for reuse across different algorithms or prims
- User-facing API is stateless and follows the plain-old data (POD) design paradigm
- Public API contains a C-Wrapper around the stateless API
- Public API contains an sklearn-like stateful wrapper around the stateless API

#### Testing

- Prims: GTests with different inputs
- Algorithms: End-to-end GTests with different inputs and different datasets

#### Documentation

- Complete and comprehensive Doxygen strings explaining the public API, restrictions, and gotchas. Any array parameters should also note whether the underlying memory is host or device. 

### Python

#### Design

> - Python class is as "near drop-in replacement" for sklearn (or relevant industry standard) API as possible

Reviewer 3: ```suggestion
- Python class is as "near drop-in replacement" for Scikit-learn (or relevant industry standard) API as possible
    - It is recommended to open an initial PR with the API design if there are going to be significant differences with reference APIs, or lack of a reference API to have a discussion about it. 
```

- Python class is pickleable
- APIs use `input_to_dev_array` to accept flexible inputs and check their datatypes

#### Testing 

> - Pytests for wrapper functionality against sklearn using relevant skleaern datasets

Reviewer 3: ```suggestion
- Pytests for wrapper functionality against Scikit-learn using relevant datasets
```

> - Stress tests against reasonable inputs (e.g short-wide, tall-narrow)

Reviewer 3: ```suggestion
- Stress tests against reasonable inputs (e.g short-wide, tall-narrow. different precision)
```

- Pytests for pickle capability

> - Pytests to evaluate correctness against sklearn on a variety of datasets

Reviewer 3: ```suggestion
- Pytests to evaluate correctness against Scikit-learn on a variety of datasets
```

- Add algorithm to benchmarks packages

#### Documentation

- Complete and comprehensive Pydoc strings explaining public API, restrictions, and gotchas

> - Docstrings include references to the underlying algorithm (e.g the paper or sklearn algo being implemented or a description of the algo used if nonstandard).

Reviewer 3: ```suggestion
- Docstrings include references to the underlying algorithm (e.g paper or Scikit-learn algorithm being implemented or a description of the algorithm used if nonstandard).
```

## Algorithm Status

Sometimes it is possible that only some of these requirements are satisfied before an algorithm gets released. We assign a status of `bronze`, `silver`, or `gold` to denote the amount of work necessary to provide the most robust, performant, and maintainable algorithm possible.

### Bronze: 

> - Known design flaws affecting sustainability and/or performance require a significant effort / redesign

Reviewer 3: ```suggestion
- Known design flaws affecting sustainability and/or performance require a significant effort / redesign / code refactor
```

> - C++ checklist is complete but Python API has not been built

Reviewer 3: ```suggestion
- C++ API functionality is complete but Python API has not been built
```

Author: I'm not sure if I agree with this one. Hoping to keep this a set of guidelines that make very clear the steps to follow in order to define things like `API functionality`. 

Perhaps we could say `C++ API functionality (see [checklist](#C++))`?

> - Python API exists without C++ API when a C++ implementation would provide a significant performance boost

Reviewer 3: ```suggestion
- Pure Python API exists without using C++ for processing when a C++ implementation would provide a significant performance boost
```

> - There are known bugs that affect memory / accuracy on some inputs

Reviewer 3: ```suggestion
- There are known issues that affect memory / accuracy / stability on some inputs
```

> - Coverage of automated tests in C++ or Python is minimal and could be significantly improved

Reviewer 3: ```suggestion
- Coverage of automated tests in C++ or Python is minimal and could be significantly improved
- Outdated / legacy code that requires significant work and refactoring
```

### Silver:
- Known design flaws affecting sustainability and/or performance are fairly low in complexity (can be fixed quickly and without too much effort)

> - Automated tests in C++ or Python are not as extensive as they could be, but they do test general functionality

Reviewer 3: ```suggestion
- Automated tests in C++ or Python are not as extensive as they could be regarding different datasets / inputs, but they have complete code coverage
```

Author: Using `have complete code coverage` seems to counter the `not extensive as they could be`. If the tests have complete code coverage, I'd say they are done. Our definitions of `complete` and `code coverage` might be different here.

- There are known (or potential) bugs that affect memory / accuracy, but only on a small subset of specific inputs
- There are known optimizations that would provide a significant performance boost over the current implementation

### Gold: 
- Design is robust, maintainable, and performant without any known flaws.
- Algorithm is as performant as possible without any significant known or potential optimizations over the current implementation.
- Algorithm is consistently correct and bug free on all datasets.
- There are no known memory usage limitations resulting from the design choices that could be reasonably refactored.

> 

Reviewer 3: I would add a note that this classification of algorithms is dynamic, for example a Gold algorithm could get downgraded to Silver due to needing to be updated to new low level APIs or libraries.

# Defining cuML's Definition of Done Criteria

## Algorithm Completion Checklist

Below is a quick and simple checklist for developers to determine whether an algorithm is complete and ready for release. Most of these items contain more detailed descriptions in their corresponding developer guide. The checklist is broken down by layer (C++ or Python) and categorized further into

- **Design:** All algorithms should be designed with an eye on maintainability, performance, readability, and robustness.

> - **Testing:** The goal for automated testing is to increase the code coverage as much as possible in order to ease time spent fixing bugs and developing new features. Additionally, a very important factor for a tool like `cuml` is to provide testing with multitple datasets that really stress the mathematical behavior of the algorithms, even if they don't increase code coverage! A comprehensive set of tests lowers the possibility for regressions and the introduction of bugs as the code evolves between versions. This covers both correctness & performance. 

Reviewer: ```suggestion
- **Testing:** The goal for automated testing is to increase the code coverage as much as possible in order to ease time spent fixing bugs and developing new features. Additionally, a very important factor for a tool like `cuml` is to provide testing with multiple datasets that really stress the mathematical behavior of the algorithms, even if they don't increase code coverage! A comprehensive set of tests lowers the possibility for regressions and the introduction of bugs as the code evolves between versions. This covers both correctness & performance. 
```
Typo with 'multiple datasets'

> - **Testing:** The goal for automated testing is to increase the code coverage as much as possible in order to ease time spent fixing bugs and developing new features. Additionally, a very important factor for a tool like `cuml` is to provide testing with multitple datasets that really stress the mathematical behavior of the algorithms, even if they don't increase code coverage! A comprehensive set of tests lowers the possibility for regressions and the introduction of bugs as the code evolves between versions. This covers both correctness & performance. 

Reviewer: Not quite sure I got the message here. Why add and run any more extra tests if they do not improve coverage? May be we should reword it to "add multiple datasets precisely to help improve coverage"?

Author: This was a change requested by @dantegd and I think he's defining coverage as a lateral, physical path through the code, rather than better covering the space of possible inputs (not necessarily taking different physical paths through the code). 

Semantics aside, I do agree with you, @teju85, that the wording `even if they don't improve code coverage` can, and should, be improved.

Author: @teju85 @dantegd I've adjusted the wording for this bullet point based on our discussion. Can you guys check the updated version and let me know if this works for you?

- **Documentation:** User-facing documentation should be complete and descriptive. Developer-facing documentation should be used for constructs which are complex and/or not immediately obvious. 

### C++

#### Design

- Existing prims are used wherever possible
- Array inputs and outputs to algorithms are accepted on device
- New prims created wherever there is potential for reuse across different algorithms or prims

> - User-facing API is stateless and follows the plain-old data (POD) design paradigm

Reviewer: ```suggestion
- User-facing API is [stateless](cpp/DEVELOPER_GUIDE.md#public-cuml-interface) and follows the [plain-old data (POD)](https://en.wikipedia.org/wiki/Passive_data_structure) design paradigm
```

- Public API contains a C-Wrapper around the stateless API
- Public API contains an Scikit-learn-like stateful wrapper around the stateless API

#### Testing

- Prims: GTests with different inputs
- Algorithms: End-to-end GTests with different inputs and different datasets

#### Documentation

> - Complete and comprehensive Doxygen strings explaining the public API, restrictions, and gotchas. Any array parameters should also note whether the underlying memory is host or device. 

Reviewer: ```suggestion
- Complete and comprehensive Doxygen strings explaining the public API, restrictions, and gotchas. Any array parameters should also note whether the underlying memory is host or device.
- Array inputs/outputs should also mention their expected size/dimension.
- If there are references to the underlying algorithm, they must be cited too.
```

### Python

#### Design

- Python class is as "near drop-in replacement" for Scikit-learn (or relevant industry standard) API as possible

>     - It is recommended to open an initial PR with the API design if there are going to be significant differences with reference APIs, or lack of a reference API, to have a discussion about it. 

Reviewer: ```suggestion
- It is recommended to open an initial PR with the API design if there are going to be significant differences with reference APIs, or lack of a reference API, to have a discussion about it. 
```
Knitpick: Fixing the indentation

- Python class is pickleable
- APIs use `input_to_dev_array` to accept flexible inputs and check their datatypes

#### Testing 

- Pytests for wrapper functionality against Scikit-learn using relevant datasets
- Stress tests against reasonable inputs (e.g short-wide, tall-narrow, different numerical precision)
- Pytests for pickle capability
- Pytests to evaluate correctness against Scikit-learn on a variety of datasets
- Add algorithm to benchmarks packages

#### Documentation

- Complete and comprehensive Pydoc strings explaining public API, restrictions, and gotchas
- Docstrings include references to the underlying algorithm (e.g paper or Scikit-learn algorithm being implemented or a description of the algorithm used if nonstandard).

## Algorithm Status

> Sometimes it is possible that only some of these requirements are satisfied before an algorithm gets released. We assign a status of `bronze`, `silver`, or `gold` to denote the amount of work necessary to provide the most robust, performant, and maintainable algorithm possible.

Reviewer 2: How about "we assign a status of `bronze`, `silver`, or `gold` to denote an algorithm's level of completeness and robustness"?

I'm thinking it's not really about the amount of work remaining.

> Sometimes it is possible that only some of these requirements are satisfied before an algorithm gets released. We assign a status of `bronze`, `silver`, or `gold` to denote the amount of work necessary to provide the most robust, performant, and maintainable algorithm possible.

Reviewer 2: Should we go through and backfill these classifications for existing algos at the beginning of 0.10? Are you thinking we should make a list on the wiki or embed in the docstring for each algo? (I'm kind of partial to including it in each algo so it's immediately obvious to someone using the code, but I could be wrong.)

Reviewer: I agree with @JohnZed. Such a classification would in-turn also help us in prioritizing the work.

How about this?
We can create a checklist based on the items described in this doc.
Depending on how many items have been checked-off from this list, we can attach a number to it.
Cumulative number could then be used to decide this classification.

Note that a change to an algorithm's classification can go up or down, depending on circumstances. For example, an algorithm that has been considered `Gold status` might drop to a `Silver status` when being updated to a new low-level feature or library. 

### Bronze: 
- Known design flaws affecting sustainability and/or performance require a significant effort / redesign / refactor
- C++ checklist / API functionality is complete but Python API has not been built

> - ure Python API exists without using C++ for processing when a C++ implementation would provide a significant performance boost

Reviewer 2: ~s/ure/Pure/`

- There are known issues that affect memory / accuracy / stability on some inputs
- Coverage of automated tests in C++ or Python is minimal and could be significantly improved
- Outdated / legacy code that requires significant work and refactoring

### Silver:

> - Known design flaws affecting sustainability and/or performance are fairly low in complexity (can be fixed quickly and without too much effort)

Reviewer 2: I personally think we should focus the levels on end-user and developer experience rather than effort needed to improve. So, a design flaw affecting stability that is very minor but requires a *huge* effort to solve could still exist in silver...

> - Automated tests in C++ or Python are not as extensive as they could be regarding different datasets / inputs, but they do test general functionality

Reviewer 2: `s/as they could be/as they should be/`
